File: jtablayout2xx/src/main/java/com/liang/widget/TabView.java
package com.liang.widget;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.LayerDrawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v4.view.ViewCompat;
import android.support.v7.content.res.AppCompatResources;
import android.support.v7.widget.TooltipCompat;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.liang.jtablayout.badge.Badge;
import com.liang.jtablayout.tab.Tab;
import com.liang.jtablayout.tab.TabChild;
import com.liang.jtablayout.ripple.RippleUtils;
import com.liang.jtablayout.utils.ColorUtils;
import com.liang.jtablayoutx.R;


public class TabView extends FrameLayout implements TabChild {

    private Tab tab;
    private TextView textView;
    private ImageView iconView;
    private View badgeView;

    @Nullable
    private Drawable baseBackgroundDrawable;
    private View tabView;

    public TabView(@NonNull Context context) {
        this(context, null);

    }

    public TabView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TabView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.setFocusable(true);
        this.setClickable(true);

    }

    private void updateLayout() {
        removeAllViews();

        tabView = setContentView();
        if (tabView == null) {
            tabView = LayoutInflater.from(getContext()).inflate(
                    tab.getInline() ? R.layout.tab_item_horizontal : R.layout.tab_item_vertical, null, true);
        }
        iconView = setTabIconView();
        textView = setTabTitleView();
        badgeView = setTabBadgeView();

        addView(tabView);
        updateBackgroundDrawable(getContext());
        update();
    }

    private View setActiveView() {
        return tabView.findViewById(R.id.tab);
    }

    protected View setContentView() {
        return null;
    }

    protected TextView setTabTitleView() {
        return tabView.findViewById(R.id.tab_title);
    }

    protected ImageView setTabIconView() {
        return tabView.findViewById(R.id.tab_icon);
    }

    @Override
    public void updateBackgroundDrawable(Context context) {
        if (tab.getTabBackgroundResId() != 0) {
            this.baseBackgroundDrawable = AppCompatResources.getDrawable(context, tab.getTabBackgroundResId());
            if (this.baseBackgroundDrawable != null && this.baseBackgroundDrawable.isStateful()) {
                this.baseBackgroundDrawable.setState(this.getDrawableState());
            }
        } else {
            this.baseBackgroundDrawable = null;
        }

        Drawable contentDrawable = new GradientDrawable();
        ((GradientDrawable) contentDrawable).setColor(0);
        Object background;
        if (tab.getTabRippleColorStateList() != null) {
            GradientDrawable maskDrawable = new GradientDrawable();
            maskDrawable.setCornerRadius(1.0E-5F);
            maskDrawable.setColor(-1);
            ColorStateList rippleColor = RippleUtils.convertToRippleDrawableColor(tab.getTabRippleColorStateList());
            if (Build.VERSION.SDK_INT >= 21) {
                background = new RippleDrawable(rippleColor, tab.isUnboundedRipple() ? null : contentDrawable, tab.isUnboundedRipple() ? null : maskDrawable);
            } else {
                Drawable rippleDrawable = DrawableCompat.wrap(maskDrawable);
                DrawableCompat.setTintList(rippleDrawable, rippleColor);
                background = new LayerDrawable(new Drawable[]{contentDrawable, rippleDrawable});
            }
        } else {
            background = contentDrawable;
        }

        ViewCompat.setBackground(this, (Drawable) background);
        tab.getParent().postInvalidate();
    }

    protected BadgeView setTabBadgeView() {
        return tabView.findViewById(R.id.tab_badgeView);
    }

    @Override
    public void drawBackground(Canvas canvas) {
        if (this.baseBackgroundDrawable != null) {
            this.baseBackgroundDrawable.setBounds(this.getLeft(), this.getTop(), this.getRight(), this.getBottom());
            this.baseBackgroundDrawable.draw(canvas);
        }
    }

    @Override
    protected void drawableStateChanged() {
        super.drawableStateChanged();
        boolean changed = false;
        int[] state = this.getDrawableState();
        if (this.baseBackgroundDrawable != null && this.baseBackgroundDrawable.isStateful()) {
            changed |= this.baseBackgroundDrawable.setState(state);
        }

        if (changed) {
            this.invalidate();
            tab.getParent().invalidate();
        }
    }

    @Override
    public boolean performClick() {
        boolean handled = super.performClick();
        if (this.tab != null) {
            if (!handled) {
                this.playSoundEffect(0);
            }

            this.tab.select();
            return true;
        } else {
            return handled;
        }
    }

    @Override
    public void setSelected(boolean selected) {
        boolean changed = this.isSelected() != selected;
        super.setSelected(selected);
        if (changed && selected && Build.VERSION.SDK_INT < 16) {
            this.sendAccessibilityEvent(4);
        }


        if (this.iconView != null) {
            this.iconView.setSelected(selected);
            Drawable normalIcon = tab != null ? tab.getNormalIcon() : null;
            Drawable selectedIcon = tab != null ? tab.getSelectedIcon() : null;
            if (normalIcon == null && selectedIcon == null) {
                this.iconView.setVisibility(GONE);
            } else {
                this.iconView.setVisibility(VISIBLE);
                this.iconView.setImageDrawable(selected ? selectedIcon != null ? selectedIcon : normalIcon : normalIcon != null ? normalIcon : selectedIcon);
            }
        }

        if (this.textView != null) {
            this.textView.setSelected(selected);
            if (tab.isTabTextBold()) {
                textView.setTypeface(Typeface.defaultFromStyle(isSelected() ? Typeface.BOLD : Typeface.NORMAL));
            }
        }
    }

    @Override
    public void onInitializeAccessibilityEvent(AccessibilityEvent event) {
        super.onInitializeAccessibilityEvent(event);
        event.setClassName(android.support.v7.app.ActionBar.Tab.class.getName());
    }

    @TargetApi(14)
    @Override
    public void onInitializeAccessibilityNodeInfo(AccessibilityNodeInfo info) {
        super.onInitializeAccessibilityNodeInfo(info);
        info.setClassName(android.support.v7.app.ActionBar.Tab.class.getName());
    }

    @Override
    public void setTab(@Nullable Tab tab) {
        if (tab != null) {
            this.tab = tab;
            updateLayout();
        }
    }

    @Override
    public void reset() {
        this.setTab(null);
        this.setSelected(false);
    }

    @Override
    public void update() {
        Tab tab = this.tab;

        if (tab == null) {
            throw new NullPointerException("Tab is null");
        }

        ViewCompat.setPaddingRelative(this, tab.getTabPaddingStart(), tab.getTabPaddingTop(), tab.getTabPaddingEnd(), tab.getTabPaddingBottom());

//        TextViewCompat.setTextAppearance(this.textView, tab.parent.tabTextAppearance);
        if (textView != null) {
            if (tab.getTextColor() != null) {
                this.textView.setTextColor(tab.getTextColor());
            }

            if (tab.getTabTextSize() > 0) {
                this.textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, tab.getTabTextSize());
            }
        }

        updateTextAndIcon(this.textView, this.iconView);

        if (!TextUtils.isEmpty(tab.getContentDesc())) {
            this.setContentDescription(tab.getContentDesc());
        }

        this.setSelected(tab.isSelected());
    }

    @Override
    public void updateOrientation(boolean inline) {
        this.updateLayout();
    }

    @Override
    public View getView() {
        return this;
    }

    private void updateTextAndIcon(@Nullable TextView textView, @Nullable ImageView iconView) {
        if (textView == null && iconView == null) {
            return;
        }

        Drawable normalIcon = tab != null ? tab.getNormalIcon() : null;
        Drawable selectedIcon = tab != null ? tab.getSelectedIcon() : null;
        Drawable icon = tab != null && tab.getIcon() != null ? DrawableCompat.wrap(tab.getIcon()).mutate() : null;

        if (iconView != null) {
            if (normalIcon == null && selectedIcon == null && icon == null) {
                iconView.setVisibility(GONE);
                iconView.setImageDrawable(null);
            } else if (normalIcon != null || selectedIcon != null) {
                iconView.setVisibility(VISIBLE);
                iconView.setImageDrawable(iconView.isSelected() ? selectedIcon != null ? selectedIcon : normalIcon : normalIcon != null ? normalIcon : selectedIcon);
                this.setVisibility(VISIBLE);
            } else {
                iconView.setVisibility(VISIBLE);
                DrawableCompat.setTintList(icon, tab.getTabIconTint());
                if (tab.getTabIconTintMode() != null) {
                    DrawableCompat.setTintMode(icon, tab.getTabIconTintMode());
                }
            }
        }

        CharSequence text = this.tab != null ? this.tab.getText() : null;
        boolean hasText = !TextUtils.isEmpty(text);

        if (textView != null) {
            if (hasText) {
                textView.setText(text);
                textView.setVisibility(VISIBLE);
                this.setVisibility(VISIBLE);
            } else {
                textView.setVisibility(GONE);
                textView.setText(null);
            }
            if (tab.isTabTextBold()) {
                textView.setTypeface(Typeface.defaultFromStyle(isSelected() ? Typeface.BOLD : Typeface.NORMAL));
            }
        }

        CharSequence contentDesc = this.tab != null ? this.tab.getContentDesc() : null;
        TooltipCompat.setTooltipText(this, hasText ? null : contentDesc);
    }

    @Override
    public int getContentWidth() {
        boolean initialized = false;
        int left = 0;
        int right = 0;
        View[] var4 = new View[]{this.textView, this.iconView};
        int var5 = var4.length;

        for (int var6 = 0; var6 < var5; ++var6) {
            View view = var4[var6];
            if (view != null && view.getVisibility() == VISIBLE) {
                left = initialized ? Math.min(left, view.getLeft()) : view.getLeft();
                right = initialized ? Math.max(right, view.getRight()) : view.getRight();
                initialized = true;
            }
        }

        return right - left;
    }

    @Override
    public void showBadge(String msg) {
        if (badgeView != null && badgeView instanceof BadgeView) {
            ((BadgeView) badgeView).show(msg);
        }
    }

    @Override
    public void hideBadge() {
        if (badgeView != null && badgeView instanceof BadgeView) {
            ((BadgeView) badgeView).hide();
        }
    }

    @Override
    public void updateColor(float offset) {
        textView.setTextColor(ColorUtils.getColorFrom(tab.getTextColor().getColorForState(EMPTY_STATE_SET, Color.GRAY),
                tab.getTextColor().getColorForState(SELECTED_STATE_SET, Color.GRAY), offset));
    }

    @Override
    public void updateScale(float scale) {
        setScaleX(scale);
        setScaleY(scale);
    }

    @Override
    public void setBadgeTextColor(int color) {
        if (badgeView != null && badgeView instanceof Badge) {
            ((Badge) badgeView).setBadgeTextColor(color);
        }
    }

    @Override
    public void setBadgeTextSize(float sp) {
        if (badgeView != null && badgeView instanceof Badge) {
            ((Badge) badgeView).setBadgeTextSize(sp);
        }
    }

    @Override
    public void setBadgeBackgroundColor(int color) {
        if (badgeView != null && badgeView instanceof Badge) {
            ((Badge) badgeView).setBadgeBackgroundColor(color);
        }
    }

    @Override
    public void setBadgeStroke(int width, int color) {
        if (badgeView != null && badgeView instanceof Badge) {
            ((Badge) badgeView).setBadgeStroke(width, color);
        }
    }

    public Tab getTab() {
        return this.tab;
    }
}

File: tablayout3x/src/main/java/com/liang/widget/TabView.kt
package com.liang.widget

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.*
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.LayerDrawable
import android.graphics.drawable.RippleDrawable
import android.os.Build
import android.support.v4.content.ContextCompat
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v4.view.PointerIconCompat
import android.support.v4.view.ViewCompat
import android.support.v7.content.res.AppCompatResources
import android.support.v7.widget.TooltipCompat
import android.text.TextUtils
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import com.liang.tablayout3x.ItemDecoration
import com.liang.tablayout3x.R
import com.liang.tablayout3x.Tab
import com.liang.utils.*

open class TabView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) :
    FrameLayout(context, attrs, defStyleAttr), Tab {


    @SuppressLint("CustomViewStyleable")
    private fun initAttrs(context: Context, attrs: AttributeSet?, defStyleAttr: Int = 0) {
        val typedArray = context.obtainStyledAttributes(
            attrs, R.styleable.TabItem,
            defStyleAttr, 0
        )
        title = typedArray.getText(R.styleable.TabItem_android_title)
        this.defaultIcon = typedArray.getDrawable(R.styleable.TabItem_android_icon)
        tabTitleSize =
            typedArray.getDimensionPixelSize(R.styleable.TabItem_android_textSize, 0).toFloat()
        tabIconTint = context.getColorStateList(
            typedArray,
            R.styleable.TabItem_android_icon
        )
        tabIconTintMode = parseTintMode(
            typedArray.getInt(R.styleable.TabItem_android_tintMode, -1),
            null
        )

        this.tabRippleColor = typedArray.getColor(
            R.styleable.TabItem_rippleColor, 0
        )
        this.tabUnboundedRipple = typedArray.getBoolean(R.styleable.TabItem_unboundedRipple, false)

        tabBackgroundResId =
            typedArray.getResourceId(R.styleable.TabItem_android_background, 0)
        if (typedArray.hasValue(R.styleable.TabItem_android_textColor)) {
            tabTitleColors = context.getColorStateList(
                typedArray,
                R.styleable.TabItem_android_textColor
            )
        }
        typedArray.recycle()
    }

    override val contentWidth: Int
        get() {
            var initialized = false
            var left = 0
            var right = 0
            val var4 = arrayOf(textView, iconView, badgeView)
            val var5 = var4.size

            for (var6 in 0 until var5) {
                val view = var4[var6]
                if (view != null && view.visibility == 0) {
                    left = if (initialized) left.coerceAtMost(view.left) else view.left
                    right = if (initialized) right.coerceAtLeast(view.right) else view.right
                    initialized = true
                }
            }

            return right - left
        }

    override var title: CharSequence? = null
        set(value) {
            field = value
            updateView()
        }

    override var defaultIcon: Drawable? = null
        set(value) {
            field = value
            updateView()
        }
    override var normalIcon: Drawable? = null
        set(value) {
            field = value
            updateView()
        }
    override var selectedIcon: Drawable? = null
        set(value) {
            field = value
            updateView()
        }

    override val view: View
        get() = this

    override var position: Int = -1

    override var tabLayout: TabLayout? = null

    override var tabTitleColors: ColorStateList? = null
        set(value) {
            field = value
            updateView()
        }
    override var tabTitleSize: Float = 0F
        set(value) {
            field = value
            updateView()
        }
    override var tabIconTint: ColorStateList? = null
        set(value) {
            field = value
            updateView()
        }

    override var tabIconTintMode: PorterDuff.Mode? = null
        set(value) {
            field = value
            updateView()
        }

    override var isTabTitleBold: Boolean = false
        set(value) {
            field = value
            updateView()
        }
    override var inlineLabel: Boolean = false
        set(value) {
            field = value
            initTabView()
        }

    override var tabUnboundedRipple: Boolean = false
        set(value) {
            field = value
            updateBackgroundDrawable()
        }
    override var tabBackgroundResId: Int = 0
        set(value) {
            field = value
            updateBackgroundDrawable()
        }
    override var tabDecoration: ItemDecoration? = null
        set(value) {
            field = value
            tabLayout?.invalidate()
        }

    override var tabRippleColor: Int = 0
        set(value) {
            field = value
            updateBackgroundDrawable()
        }


    override fun setTitle(title: String): Tab {
        this.title = title
        return this
    }

    override fun setTitleColor(normalColor: Int, selectedColor: Int): Tab {
        return setTitleColor(createColorStateList(normalColor, selectedColor))
    }

    override fun setTitleColor(textColor: ColorStateList?): Tab {
        this.tabTitleColors = textColor
        return this
    }

    override fun setTitleSize(sizePx: Float): Tab {
        this.tabTitleSize = sizePx
        return this
    }

    override fun setTitleBold(isBold: Boolean): Tab {
        this.isTabTitleBold = isBold
        return this
    }

    override fun setIcon(icon: Int): Tab {
        val drawable: Drawable? = ContextCompat.getDrawable(context, icon)
        return setIcon(drawable)
    }

    override fun setIcon(icon: Drawable?): Tab {
        this.defaultIcon = icon
        return this
    }

    override fun setIcon(normalIcon: Int, selectedIcon: Int): Tab {
        val normalDrawable: Drawable? = ContextCompat.getDrawable(this.context, normalIcon)
        val selectedDrawable: Drawable? = ContextCompat.getDrawable(this.context, selectedIcon)
        return setIcon(normalDrawable, selectedDrawable)
    }

    override fun setIcon(normalIcon: Drawable?, selectedIcon: Drawable?): Tab {
        this.normalIcon = normalIcon
        this.selectedIcon = selectedIcon
        return this
    }

    override fun setIconTint(normalColor: Int, selectedColor: Int): Tab {
        return setIconTint(createColorStateList(normalColor, selectedColor))
    }

    override fun setIconTint(tabIconTint: ColorStateList): Tab {
        this.tabIconTint = tabIconTint
        return this
    }

    override fun setRippleColor(tabRippleColor: Int): Tab {
        this.tabRippleColor = tabRippleColor
        return this
    }

    override fun setBackground(resId: Int): Tab {
        this.tabBackgroundResId = resId
        return this
    }

    override fun setUnboundedRipple(unboundedRipple: Boolean): Tab {
        this.tabUnboundedRipple = unboundedRipple
        return this
    }

    override fun setDecoration(itemDecoration: ItemDecoration): Tab {
        this.tabDecoration = itemDecoration
        return this
    }

    override fun setTabPadding(
        tabPaddingStart: Int,
        tabPaddingTop: Int,
        tabPaddingEnd: Int,
        tabPaddingBottom: Int
    ): Tab {
        ViewCompat.setPaddingRelative(
            this,
            tabPaddingStart,
            tabPaddingTop,
            tabPaddingEnd,
            tabPaddingBottom
        )
        return this
    }

    private lateinit var tabView: View
    private var textView: TextView? = null
    private var iconView: ImageView? = null
    private var badgeView: BadgeView? = null

    private val params = LayoutParams(
        LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT
    )

    override fun select() {
        requireNotNull(tabLayout) { "Tab not attached to a TabLayout" }
        tabLayout?.selectTab(this)
    }

    override fun reset() {
        tabLayout = null
        tag = null
        defaultIcon = null
        normalIcon = null
        selectedIcon = null
        title = null
        position = -1
        tabTitleColors = null
        tabIconTintMode = null
        tabIconTint = null
        tabTitleSize = 0f
        tabRippleColor = 0
        tabBackgroundResId = 0
        this.isSelected = false
    }

    override fun showBadge(msg: String?) {
        badgeView?.show(msg)
    }

    override fun hideBadge() {
        badgeView?.hide()
    }

    override fun setBadgeTextColor(color: Int) {
        badgeView?.setBadgeTextColor(color)
    }

    override fun setBadgeTextSize(sp: Float) {
        badgeView?.setBadgeTextSize(sp)
    }

    override fun setBadgeBackgroundColor(color: Int) {
        badgeView?.setBadgeBackgroundColor(color)
    }

    override fun setBadgeStroke(width: Int, color: Int) {
        badgeView?.setBadgeStroke(width, color)
    }

    override fun updateColor(offset: Float) {
//        Log.e("Item", "Item: ${position},updateColor: $offset")
        tabTitleColors?.let {
            textView?.setTextColor(
                getColorFrom(
                    it.getColorForState(
                        View.EMPTY_STATE_SET,
                        Color.GRAY
                    ), it.getColorForState(
                        View.SELECTED_STATE_SET,
                        Color.GRAY
                    ), offset
                )
            )
        }
    }

    override fun updateScale(scale: Float) {
        scaleX = scale
        scaleY = scale
    }

    init {
        this.isClickable = true
        initTabView()
        ViewCompat.setPointerIcon(this, PointerIconCompat.getSystemIcon(context, 1002))
    }

    private fun initTabView() {
        removeAllViews()
        tabView = setContentView()
        iconView = setTabIconView()
        textView = setTabTitleView()
        badgeView = setTabBadgeView()
        params.gravity = Gravity.CENTER
        addView(tabView, params)
        updateBackgroundDrawable()
        updateView()
    }

    protected open fun setContentView(): View = LayoutInflater.from(context).inflate(
        if (inlineLabel) R.layout.tab_item_horizontal else R.layout.tab_item_vertical, null
    )

    protected open fun setTabTitleView(): TextView? = tabView.findViewById(R.id.tab_title)

    protected open fun setTabIconView(): ImageView? = tabView.findViewById(R.id.tab_icon)

    protected open fun setTabBadgeView(): BadgeView? = tabView.findViewById(R.id.tab_badgeView)

    override fun performClick(): Boolean {
        val handled = super.performClick()
        return if (tabLayout != null) {
            if (!handled) {
                playSoundEffect(0)
            }
            select()
            true
        } else {
            handled
        }
    }

    private fun updateView() {
        updateTextAndIcon(textView, iconView)
        if (!TextUtils.isEmpty(contentDescription)) {
            this.contentDescription = contentDescription
        }
        this.isSelected = isSelected
    }

    private fun updateTextAndIcon(textView: TextView?, iconView: ImageView?) {
        iconView?.let { iv ->
            val icon =
                if (defaultIcon != null) DrawableCompat.wrap(defaultIcon!!).mutate() else null
            if (normalIcon == null && selectedIcon == null && icon == null) {
                iv.visibility = GONE
                iv.setImageDrawable(null)
            } else if (normalIcon != null || selectedIcon != null) {
                iv.visibility = VISIBLE
                iv.setImageDrawable(
                    if (iv.isSelected) {
                        selectedIcon ?: normalIcon
                    } else {
                        normalIcon
                    }
                )
                this.visibility = VISIBLE
            } else {
                iv.visibility = VISIBLE
                icon?.let { ic ->
                    tabIconTint?.let {
                        DrawableCompat.setTintList(ic, it)
                    }
                    tabIconTintMode?.let {
                        DrawableCompat.setTintMode(ic, it)
                    }
                }
                iv.setImageDrawable(icon)
            }
        }

        val hasText = !title.isNullOrEmpty()

        textView?.let { tv ->
            if (hasText) {
                tv.text = title
                tv.visibility = VISIBLE
                this.visibility = VISIBLE
            } else {
                tv.visibility = GONE
                tv.text = null
            }

            tabTitleColors?.let {
                tv.setTextColor(it)
            }

            if (tabTitleSize > 0f) {
                tv.setTextSize(TypedValue.COMPLEX_UNIT_PX, tabTitleSize)
            }

            if (isTabTitleBold) {
                tv.typeface =
                    Typeface.defaultFromStyle(if (isSelected) Typeface.BOLD else Typeface.NORMAL)
            }
        }

        val contentDesc = contentDescription
        TooltipCompat.setTooltipText(this, if (hasText) null else contentDesc)
    }

    override fun setSelected(selected: Boolean) {
        val changed = this.isSelected != selected
        super.setSelected(selected)
        if (changed && selected && Build.VERSION.SDK_INT < 16) {
            sendAccessibilityEvent(4)
        }

        iconView?.let { iv ->
            val icon =
                if (defaultIcon != null) DrawableCompat.wrap(defaultIcon!!).mutate() else null
            iv.isSelected = selected
            if (normalIcon == null && selectedIcon == null && icon == null) {
                iv.visibility = GONE
            } else {
                iv.visibility = VISIBLE
                if (normalIcon != null || selectedIcon != null) {
                    iv.setImageDrawable(
                        if (iv.isSelected) {
                            selectedIcon ?: normalIcon
                        } else {
                            normalIcon
                        }
                    )
                }
            }
        }

        this.textView?.isSelected = selected
        if (isTabTitleBold) {
            textView?.typeface =
                Typeface.defaultFromStyle(if (isSelected) Typeface.BOLD else Typeface.NORMAL)
        }
    }

    private var baseBackgroundDrawable: Drawable? = null

    private fun updateBackgroundDrawable() {
        if (tabBackgroundResId != 0) {
            this.baseBackgroundDrawable =
                AppCompatResources.getDrawable(context, tabBackgroundResId)
            if (this.baseBackgroundDrawable != null && this.baseBackgroundDrawable?.isStateful!!) {
                this.baseBackgroundDrawable?.state = this.drawableState
            }
        } else {
            this.baseBackgroundDrawable = null
        }
        val contentDrawable: Drawable = GradientDrawable()
        (contentDrawable as GradientDrawable).setColor(0)
        val background: Any
        val tabRippleColorStateList = createColorStateList(tabRippleColor, tabRippleColor)
        val maskDrawable = GradientDrawable()
        maskDrawable.cornerRadius = 1.0E-5f
        maskDrawable.setColor(-1)
        val rippleColor = RippleUtils.convertToRippleDrawableColor(tabRippleColorStateList)
        background = if (Build.VERSION.SDK_INT >= 21) {
            RippleDrawable(
                rippleColor,
                if (this.tabUnboundedRipple) null else contentDrawable,
                if (this.tabUnboundedRipple) null else maskDrawable
            )
        } else {
            val rippleDrawable = DrawableCompat.wrap(maskDrawable)
            DrawableCompat.setTintList(rippleDrawable, rippleColor)
            LayerDrawable(arrayOf(contentDrawable, rippleDrawable))
        }
        ViewCompat.setBackground(this, background as Drawable)
        tabLayout?.invalidate()
    }

    override fun drawBackground(canvas: Canvas) {
        if (this.baseBackgroundDrawable != null) {
            this.baseBackgroundDrawable?.setBounds(
                this.left,
                this.top,
                this.right,
                this.bottom
            )
            this.baseBackgroundDrawable?.draw(canvas)
        }
    }

    override fun drawableStateChanged() {
        super.drawableStateChanged()
        var changed = false
        val state = this.drawableState
        if (this.baseBackgroundDrawable != null && this.baseBackgroundDrawable?.isStateful!!) {
            changed = changed or this.baseBackgroundDrawable?.setState(state)!!
        }
        if (changed) {
            this.invalidate()
            tabLayout?.invalidate()
        }
    }


    class DefItemDecoration @JvmOverloads constructor(
        width: Int = 0,
        height: Int = 0
    ) :
        ItemDecoration(width, height) {

        var color: Int = 0
            set(value) {
                if (field == value) {
                    return
                }
                field = value
                paint.color = color
            }

        init {
            Log.e("DefItemDecoration", "width: $width")
            paint.strokeWidth = width.toFloat()
        }

        override fun getItemOffsets(outRect: Rect) {
            outRect.set(0, 0, width, height)
        }

        override fun onDraw(canvas: Canvas, rect: Rect) {
            canvas.drawRect(rect, paint)
        }
    }

    init {
        initAttrs(context, attrs, defStyleAttr)
    }

}